Simple distributed optimizer

aazzolini opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28948 Allow to create local RRef with value
* **#28910 Simple distributed optimizer**

Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18230877](https://our.internmc.facebook.com/intern/diff/D18230877/)

aazzolini committed:
Simple distributed optimizer

Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18230877](https://our.internmc.facebook.com/intern/diff/D18230877/)

[ghstack-poisoned]

aazzolini commented:
@pytorchbot retest this please

pritamdamania87 commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
+            to optimize.
+        args: arguments to pass to the optimizer constructor on each worker.
+        kwargs: arguments to pass to the optimizer constructor on each worker.
```
Why don't we pass an instance of the optimizer instead of the class and its args? We can create an RRef of this instance on the remote node by just passing it in as a parameter to a `remote` call and just returning the same object back.

aazzolini replied:
This won't be possible -- optimizer objects (both torch.optim and the FunctionalOptimizer that I introduced above) take local model parameters as input to the constructor. However, we don't have access to those parameters from here. The worker where the parameters live is the only one capable of passing those parameters to the constructor of the optimizer.

Alternativelly we could introduce an OptimizerConfig class but it wouldn't solve the underlying issue.

pritamdamania87 commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+            self.remote_optimizers.append(remote_optim)
+
+
+    def step(self, autograd_ctx_id):
```
Should we keep this API inline with `backward` where we implicitly use the current context id instead of passing it in?

aazzolini replied:
The issue is that dist_autograd._current_context() is private. Should we expose it publicly? (In a subsequent PR).

pritamdamania87 replied:
Can we keep it private and still call it from here? There is no reason to expose this method to end users at the moment.

pritamdamania87 commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+        try:
+            fut.wait()
+        except Exception as e:
+            exception = e
```
What is the reason that we continue waiting for other futures if we see an exception? Shouldn't we just exit as soon as possible and not wait for other futures?

aazzolini replied:
I'm just not sure what would happen if one of these operations is still executing remotely and we exit the current autograd context.

pritamdamania87 commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+
+
+def _wait_for_all(rpc_futs):
+    # TODO: improve error propagation
```
What sort of improvements do we have in mind here? Also, can we create a github issue for this so we can keep track of this for the future?

aazzolini replied:
Ideally we want to gather all exceptions in a list. I'll open an issue for it.

aazzolini committed:
Update on "Simple distributed optimizer"


Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18230877](https://our.internmc.facebook.com/intern/diff/D18230877/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 3304db2:
* 1/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=3304db2ea10bb2bfab7654e2e0e790de403891ab).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

mrshenli commented on test/dist_optimizer_test.py:
```diff
@@ -0,0 +1,180 @@
+from dist_utils import INIT_METHOD_TEMPLATE, dist_init
+import torch
+import torch.distributed.autograd as dist_autograd
+import torch.distributed.optimizer.dist_optimizer as dist_optimizer
```
The corresponding package and class is called `torch.optim.Optimizer`, do we want to keep the same convention here? i.e., `torch.distributed.optim.Optimizer`.

mrshenli commented on test/dist_optimizer_test.py:
```diff
@@ -0,0 +1,180 @@
+
+
+@unittest.skipIf(
+    not torch._six.PY3, "Pytorch distributed autograd package " "does not support python2"
```
two unnecessary double quotes, or maybe you were intended to break them into shorter lines.

mrshenli commented on test/dist_optimizer_test.py:
```diff
@@ -0,0 +1,180 @@
+        remote_param1 = remote_meth(MyModule.get_w, remote_module1)
+        remote_param2 = remote_meth(MyModule.get_w, remote_module2)
+
+        dst_optim = dist_optimizer.DistributedOptimizer(
```
`dst_optim` -> `dist_optim`?

mrshenli commented on test/dist_optimizer_test.py:
```diff
@@ -0,0 +1,180 @@
+            t2 = torch.rand((3, 3), requires_grad=True)
+            output1 = rpc_async_meth(MyModule.forward, remote_module1, t2)
+            output2 = rpc_async_meth(
+                MyModule.forward, remote_module2, output1.wait())
```
In followup PRs, we also want to test passing an `RRef` of `output1` to `remote_module2`, right?

aazzolini replied:
Correct, I was waiting for the .to_here() autograd propagation PR to be landed. If it's already the case, I can add the test to this PR.

mrshenli replied:
It's landed, but we can still add new tests in a PR on top of this. There are some flakiness that we are investigating, which should not block us from landing this PR in its current form.

mrshenli commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+    def __init__(self, params):
+        self.params = params
+
+    def step(self, gradients):
```
Any reason for directly taking the `gradients` here instead of a distributed autograd context id, and use that id  to retrieve gradients?

aazzolini replied:
Notice that this class is a local optimizer, it doesn't know about distributed anything.

mrshenli commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+
+from collections import defaultdict
+
+class FunctionalOptimizer:
```
Would I be correct if I assume there is no easy way to reuse existing optimizers without modifying them as they directly reads from `param.grad`?

aazzolini replied:
I am thinking of writing an adapter that allows to use existing optim.Optimizer. In order to avoid race conditions I'll have to create multiple instance of optim.Optimizer taking the same parameters but unsharing the gradients. I can open an issue explaining the idea in more details.

mrshenli commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+    specific parameters.
+
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
```
> Can the constructor take a torch.optim class as input instead?

I also would prefer that way.

> hogwild case there could be a race in the proposal above (I'm not sure if that matters though).

This probably won't work in hogwild without a lock as grad might be overwritten by other threads, but as hogwild is a more advanced use case, it seems reasonable to ask hogwild applications to implement their own local optimizers that do not directly read from `param.grad`? We could have a mode flag here to toggle whether we want to use a lock.

aazzolini replied:
The idea I'm going to propose is to have an adapter that can take a regular optim.Optimizer and make it into a FunctionalOptimizer. There's two possible implementations for this wrapper:
1) simple wrapper: uses the lock as you suggested above;
2) hogwild-capable wrapper: maintains multiple instances of the regular optimizer sharing parameters but unshare gradients.

We can then optionally automatically wrap the Optimizer from the constructor of DistributedOptimizer so it's transparent for the user, but I'd rather not make it transparent -- because the user actually has to make a choice whether they want to block or make it hogwild.

For internal use cases, we may prefer to go to FunctionalOptimizer directly to avoid overhead of keeping multiple optimizer instances for different hogwild threads.

pritamdamania87 replied:
> avoid overhead of keeping multiple optimizer instances for different hogwild threads.

I'm not sure I completely followed this, why do we need multiple optimizer instances for different threads?

pritamdamania87 replied:
Overall I do like this idea of the adapter, although would it be possible to add the adapter as part of this PR? That way we can get a good sense of the complete API we're providing to users for this.

mrshenli commented on torch/distributed/optimizer/dist_optimizer.py:
```diff
@@ -0,0 +1,126 @@
+    Args:
+        optimizer_class (FunctionalOptimizer): the class of optimizer to
+            instantiate on each worker.
+        params_rref (list[RRef]): list of RRefs to local or remote parameters
```
Would I be correct if I assume that the `RemoteModule` we discussed offline should have a `parameters()` API that returns a list of RRefs?

aazzolini replied:
Correct. I didn't want to introduce a full fledged RemoteModule in this diff; also we'll need a prep PR before we can return a list of RRefs.

aazzolini committed:
Update on "Simple distributed optimizer"


Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18230877](https://our.internmc.facebook.com/intern/diff/D18230877/)

[ghstack-poisoned]

aazzolini commented on test/dist_optimizer_test.py:
```diff
@@ -0,0 +1,180 @@
+
+    @dist_init()
+    def test_dist_optim(self):
+        if self.rank != 0:
```
No specific reason, will remove.

ezyang commented:
Hi I was tagged reviewer here. It looks like there are already reviews. Is there any specific you want me to look at?

pritamdamania87 commented:
@ezyang I tagged you as a reviewer since this is somewhat related to distributed autograd, although looks like @vincentqb @soumith are the owners for torch.optim based on this: https://pytorch.org/docs/stable/community/persons_of_interest.html#torch-optim

aazzolini committed:
Update on "Simple distributed optimizer"


Implements a simple python distributed optimizer that takes rrefs to parameters that will be optimized.
It keeps instances of optimizers remotely and calling step on distributed optimizer will call step on each of the remote optimizers in parallel.

Differential Revision: [D18230877](https://our.internmc.facebook.com/intern/diff/D18230877/)

[ghstack-poisoned]

ezyang commented:
Yes, why don't you tag @vincentqb for review her, for optimizer notes.

mrshenli commented:
Hey @vincentqb, could you please help to take a look at the optimizer API in this PR?

vincentqb commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,112 @@
+    def __init__(self, params):
+        self.params = params
+
+    def step(self, gradients):
```
The current `Optimizer` takes and returns the loss function. What would be an equivalent here?

aazzolini replied:
Unfortunately there's no straightforward way of porting the closure() functionality to distributed as it would potentially require algorithm-specific synchronization across workers. This would be true for L-FBGS for example, where closure() and parameter updates are interleaved in a loop inside of step().

vincentqb commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,112 @@
+                       matters as the list of gradients passed to the step
+                       function must be aligned with this list.
+    """
+    def __init__(self, params):
```
Would having "defaults" make sense here to mimic `Optimizer`?

vincentqb commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,112 @@
+            self.remote_optimizers.append(remote_optim)
+
+
+    def step(self, autograd_ctx_id):
```
Same comment as `FunctionalOptimizer.step`

vincentqb commented on torch/distributed/optim/optimizer.py:
```diff
@@ -0,0 +1,112 @@
+
+from collections import defaultdict
+
+class FunctionalOptimizer:
```
Is this precise enough as a name? Or could we name it differently, say `LocalOptimizer` or `NonLockingOptimizer`?

vincentqb commented:
> Can the constructor take a torch.optim class as input instead?

What was the conclusion of this discussion? 

As mentioned above, it would be really nice to have a mechanism to re-use the current optimizers. Wrappers to offer {,non-}locking could be nice, but a little heavy to use. Offering a default through a toggle may be a solution if we don't expect more toggles to appear with time.

Also, if possible to keep the syntax as close as possible, it'd be nice to have a way to simply replace `opt = torch.optim.SGD(...)` by `opt = torch.distributed.optim.FunctionalSGD(...)` to get a FunctionalOptimizer version of an Optimizer. I highlighted some differences in the parameters passed.

Thoughts?

aazzolini commented:
Proposal:

- I'll rename FunctionalOptimizer to LocalOptimizer
- I'll keep LocalOptimizer as simple as it is right now, in particular:
  - I won't add "defaults" as argument constructor as 1) it would require code to be added to the base class but I'd prefer to keep it interface-only; 2) it would be unclear how to wrap an existing Optimizer as a LocalOptimizer since we'd have 2 instances of the "defaults" field. LocalOptimizer interface-only makes it cleaner to implement as a wrapper;
 - I won't add the "closure" argument to step() because there's no sensible default way of implementing this functionality for distributed optimizers.
- I'll implement a thin "LockingOptimizer" wrapper that will wrap a optim.Optimizer as a LocalOptimizer. If a optim.Optimzier is passed to DistributedOptimizer it will be automatically wrapped.
- I won't re-implement any specific Optimizer on top of LocalOptimizer (except in test cases).

Let me know if that works.

aazzolini commented:
@vincentqb  could you comment on this proposal since we'd like to have this PR landed soon for 1.4 release?

aazzolini commented:
Synced up offline with @mrshenli -- instead i'll go the simpler route of integrating with "optim.Optimizer" directly to avoid introducing new APIs.

vincentqb commented:
@aazzolini -- did you mean to close this PR?

> @vincentqb could you comment on this proposal since we'd like to have this PR landed soon for 1.4 release?

The proposal, with the amendment, sounds good to me. Thanks for working on this!

aazzolini commented:
This is weird, I must have closed this by mistake. It won't let me re-open it mentioning the commits have already been merged but that doesn't seem to be the case.
I'll open a new PR.

aazzolini commented:
@vincentqb @pritamdamania87  @mrshenli  I produced https://github.com/pytorch/pytorch/pull/29304 to continue the discussion.
